e2e/storage: delay CreateDriver, simplify APIs and test definition

pohly opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/release.md#issue-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
6. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:

CreateDriver is a potentially expensive operation, depending on the
driver. Creating and tearing down a framework instance also takes
time (measured at 6 seconds on a fast machine) and produces quite a
bit of log output.
    
Both can be avoided for tests that skipped based on static
information (like for instance the current OS, vendor, driver and test
pattern) by making the test suite responsible for creating framework
and driver.
    
The lifecycle of the TestConfig instance was confusing because it was
stored inside the DriverInfo, a struct which conceptually is static,
while the TestConfig is dynamic. It is cleaner to separate the two,
even if that means that an additional pointer must be passed into some
functions. Now CreateDriver is responsible for initializing the
PerTestConfig that is to be used by the test.
    
To make this approach simpler to implement (= less functions which
need the pointer) and the tests easier to read, the entire setup and
test definition is now contained in a single function. This is how it
is normally done in Ginkgo. This is easier to read because one can see
at a glance where variables are set, instead of having to trace values
though two additional structs (TestResource and TestInput).
    
Because we are changing the API already, also other changes are made:
- some function prototypes get simplified
- the naming of functions is changed to match their purpose
  (tests aren't executed by the test suite, they only get defined
  for later execution)
- unused methods get removed (TestSuite.skipUnsupportedTest is redundant)

**Which issue(s) this PR fixes**:
Fixes #72288

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
<!--  
If no, just write "NONE".
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
-->
```release-note
e2e storage tests run faster and are easier to read
```

pohly committed:
e2e/storage: introduce TestVolume interface

This increases type safety and makes the code easier to read because
it becomes obvious that the "test resource" passed to some functions
must be the result of a previous CreateVolume.

This makes it possible to remove:
- functions that never did anything (the DeleteVolume methods in
  drivers that never create a volume)
- type casts (in the DeleteVolume implementation)
- the unused DeleteVolume parameters
- the stand-alone DeleteVolume function (which would be just a non-nil
  check)

GetPersistentVolumeSource and GetVolumeSource could also become
methods on more specific interfaces - they don't actually use anything
from TestDriver instance which provides them.

The main motivation however is to reduce the number of methods which
might need an explicit test config parameter.

pohly committed:
e2e/storage: speed up skipping, simplify APIs and test definition

CreateDriver (now called SetupTest) is a potentially expensive
operation, depending on the driver. Creating and tearing down a
framework instance also takes time (measured at 6 seconds on a fast
machine) and produces quite a bit of log output.

Both can be avoided for tests that skip based on static
information (like for instance the current OS, vendor, driver and test
pattern) by making the test suite responsible for creating framework
and driver.

The lifecycle of the TestConfig instance was confusing because it was
stored inside the DriverInfo, a struct which conceptually is static,
while the TestConfig is dynamic. It is cleaner to separate the two,
even if that means that an additional pointer must be passed into some
functions. Now CreateDriver is responsible for initializing the
PerTestConfig that is to be used by the test.

To make this approach simpler to implement (= less functions which
need the pointer) and the tests easier to read, the entire setup and
test definition is now contained in a single function. This is how it
is normally done in Ginkgo. This is easier to read because one can see
at a glance where variables are set, instead of having to trace values
though two additional structs (TestResource and TestInput).

Because we are changing the API already, also other changes are made:
- some function prototypes get simplified
- the naming of functions is changed to match their purpose
  (tests aren't executed by the test suite, they only get defined
  for later execution)
- unused methods get removed (TestSuite.skipUnsupportedTest is redundant)

msau42 commented:
/assign @verult 
cc @mkimuram

pohly commented:
/cc @mkimuram

pohly commented:
This WIP because only the provisioning testsuite has been updated. To get the code compiled, the rest of the testsuites were removed.

pohly commented:
I'd like to get feedback on this approach and a decision on https://github.com/kubernetes/kubernetes/pull/70992 before continuing with this PR.

davidz627 commented:
/cc

davidz627 commented on test/e2e/framework/specs.go:
```diff
@@ -0,0 +1,47 @@
+	"github.com/onsi/ginkgo"
+)
+
+type spec struct {
```
I would prefer a more descriptive name for this struct, `spec` is an incredibly overused term and its unclear what type of spec this is

pohly replied:
`spec` is the term used by Ginkgo. But I agree, that context should be more explicit.

How about `ginkgoSpec`?

davidz627 commented on test/e2e/storage/csi_volumes.go:
```diff
@@ -153,17 +149,18 @@ var _ = utils.SIGDescribe("CSI Volumes", func() {
 			cs = f.ClientSet
 			csics = f.CSIClientSet
-			config := testsuites.TestConfig{
+			config = &testsuites.TestConfig{
```
`perTestConfig`? ditto everywhere else. I think it needs to be really clear what the lifecycle of these objects are, thats what bit us in the butt earlier. The more self-documenting the better. But we should be even clearer still by adding a comment to each of these variables mentioning when they are created/destroyed and when/if they should be modified/cleared.

pohly replied:
Okay.

davidz627 commented on test/e2e/storage/drivers/csi.go:
```diff
@@ -197,14 +190,8 @@ func (g *gcePDCSIDriver) GetDriverInfo() *testsuites.DriverInfo {
 	return &g.driverInfo
 }
 
-func (g *gcePDCSIDriver) SkipUnsupportedTest(pattern testpatterns.TestPattern) {
```
curious as to the reason why we're removing `SkipUnsupportedTest` and moving the skip logic into `CreateDriver()`?

pohly replied:
This is because this PR is based on https://github.com/kubernetes/kubernetes/pull/70992. When looking at overall changed files, you'll also see changes from commits in that PR, like this one.

Better have a look at the recent commits that are exclusive to this PR ("introduce TestVolume interface", "use local TestConfig, remove indirection").

Sorry for the confusion. I have multiple PRs pending right now that all conflict with each other, therefore I am working on a single branch :-/

davidz627 replied:
Ah I see, I will look at specific commits for future review

davidz627 commented on test/e2e/storage/testsuites/provisioning.go:
```diff
@@ -73,132 +80,81 @@ func (p *provisioningTestSuite) getTestSuiteInfo() TestSuiteInfo {
+			framework.Logf("In creating storage class object and pvc object for driver - sc: %v, pvc: %v", sc, pvc)
+			testCase = StorageClassTest{
+				Client:       config.Framework.ClientSet,
+				Claim:        pvc,
```
if we have the entire claim in this struct I there some of the other fields in this struct become redundant. e.g. claimsize, expectedsize

pohly replied:
Agreed. Not sure whether I should improve that in this PR, though.

davidz627 replied:
since this PR is adding the new fields it seems in scope to me to also remove the "old" redundant fields. Otherwise the cleanup might never happen

davidz627 commented on test/e2e/storage/testsuites/provisioning.go:
```diff
@@ -73,132 +80,81 @@ func (p *provisioningTestSuite) getTestSuiteInfo() TestSuiteInfo {
+			if sc == nil {
+				framework.Skipf("Driver %q does not define Dynamic Provision StorageClass - skipping", driver.GetDriverInfo().Name)
+			}
+			pvc = getClaim(claimSize, config.Framework.Namespace.Name)
```
nit: still think this stuff should go in a function. Maybe not `setupResource` exactly but `setupStorageClassTest` or something?

davidz627 commented on test/e2e/storage/testsuites/provisioning.go:
```diff
@@ -265,6 +221,26 @@ func TestDynamicProvisioning(t StorageClassTest, client clientset.Interface, cla
 	}
 
+	// Determine where to run pods.
+	nodeName := t.NodeName
```
is this a framework change or a test change? It's unclear to me if this belongs in the same PR

pohly replied:
Same problem: this change is from PR #72002

davidz627 commented on test/e2e/storage/testsuites/provisioning.go:
```diff
@@ -280,11 +256,53 @@ func TestDynamicProvisioning(t StorageClassTest, client clientset.Interface, cla
+		runInPodWithVolume(client, claim.Namespace, claim.Name, "-second", nodeName, "grep 'hello world' /mnt/test/data")
+
+		// Run on another node, if we have one.
+		var secondNodeName string
```
test change or framework change? Unclear if it should be in the same PR

davidz627 commented on test/e2e/storage/testsuites/testdriver.go:
```diff
@@ -70,25 +87,28 @@ type DynamicPVTestDriver interface {
 	CapBlock       Capability = "block"       // raw block mode
 	CapFsGroup     Capability = "fsGroup"     // volume ownership via fsGroup
 	CapExec        Capability = "exec"        // exec a file in the volume
+	CapMultiPODs   Capability = "multipods"   // multiple pods on a node can use the same volume concurrently; enabled by default
```
this seems out of scope

davidz627 reviewed (commented):
just a quick first pass. Thanks for working on this!

wongma7 commented on test/e2e/storage/testsuites/testdriver.go:
```diff
@@ -31,7 +31,7 @@ type TestDriver interface {
 	// except CreateVolume. May be called more than once and should only do something on
 	// the first call.
-	CreateDriver()
+	CreateDriver(config *TestConfig)
```
I think it would be clearer to have CreateDriver return a config and not take a pointer, instead of modifying the pointer argument. Init*Driver returns a static DriverInfo so CreateDriver should similarly return a dynamic TestConfig.

wongma7 replied:
since most of the implementations are using config.framework a bunch maybe pass framework and prefix instead and let CreateDriver construct the TestConfig in full

pohly replied:
It's true that the framework pointer could also be passed in earlier. However, it is unusable at init time because the framework initializes the content in a `BeforeEach`. I think it is cleaner to only grant access to the pointer when it is really initialized and needed, and not sooner.

pohly replied:
I ended up letting `CreateDriver` choose the prefix. This is a slight change in behavior, but once one gets used to it, it might even be better because we get different prefixes into the test output.

wongma7 commented on test/e2e/storage/testsuites/testdriver.go:
```diff
@@ -59,7 +59,7 @@ type TestVolume interface {
 	TestDriver
 	// CreateVolume creates a pre-provisioned volume of the desired volume type.
-	CreateVolume(volumeType testpatterns.TestVolType) TestVolume
+	CreateVolume(config *TestConfig, volumeType testpatterns.TestVolType) TestVolume
```
same as with CreateDriver, I think it would be clearer to return a TestConfig if CreateVolume may modify the config

wongma7 replied:
yes

pohly replied:
Not done yet.

mathu97 commented on test/e2e/storage/testsuites/base.go:
```diff
@@ -232,7 +232,9 @@ func (r *genericVolumeTestResource) cleanupResource(driver TestDriver, pattern t
 	// Cleanup volume for pre-provisioned volume tests
-	DeleteVolume(driver, volType, r.driverTestResource)
+	if r.volume != nil {
+		r.volume.DeleteVolume()
```
Should `r.volume` be set to `nil` after this?

pohly replied:
`r.volume` is still valid and `r` goes out of scope soon after. I personally wouldn't bother.

wongma7 commented on test/e2e/storage/testsuites/provisioning.go:
```diff
@@ -73,132 +80,81 @@ func (p *provisioningTestSuite) getTestSuiteInfo() TestSuiteInfo {
-			// Create test input
-			resource, input = createProvisioningTestInput(driver, pattern)
-		})
+		Expect(pattern.VolType).To(Equal(testpatterns.DynamicPV))
```
Expect is outside If

why do we want to rename RunTestSuite and execTest to setup*?

pohly replied:
Right, this `pattern.VolType` assertion must be done differently.

`RunTestSuite` and `execTest` are misnamed because they don't actual run or execute tests. They just define the test suite. The tests themselves then run later, when Ginkgo starts the test suite.

This is a subtle difference, but it is something that test developers have to understand.

wongma7 replied:
:+1: gotcha...an alternative name to "setup" could be "specify" or "define" as you've used in the comments for these functions. I took "setup" to mean purely the kind of "setting up" in `BeforeEach`es and went looking for where the `It` ended up for a bit. Might just be me...

pohly replied:
It changed it to use `define`.

pohly commented:
Big force-push. The revised commits are now based on master, because https://github.com/kubernetes/kubernetes/pull/70992 will not be merged and https://github.com/kubernetes/kubernetes/pull/72002 can be handled separately.

Instead of skipping tests at definition time (https://github.com/kubernetes/kubernetes/pull/70992), I ended up moving the expensive operations in this PR behind most of the runtime skip checks - that also makes it faster to run the entire test suite.

@wongma7 @davidz627 @mathu97 please have another look.

pohly commented:
@mkimuram what was the rationale for introducing the "tune patterns" callback? It looks like it is used by CSI to filter out certain unsupported test cases based on the patterns:

https://github.com/kubernetes/kubernetes/blob/33a9c6e892f69e20be9527ba00bd33dfa5de221b/test/e2e/storage/csi_volumes.go#L66-L78

When I tried to extend this filtering to more test cases (https://github.com/kubernetes/kubernetes/pull/70992), it was pointed out that this makes it hard to discover why a test isn't executed and that runtime skipping is the preferred approach for tests that cannot run.

Taking that into account, should this "tune patterns" filtering perhaps be removed? This PR revamps the API anyway, so I could simply do it here.

wongma7 commented on test/e2e/storage/drivers/csi.go:
```diff
@@ -310,8 +306,11 @@ func (g *gcePDCSIDriver) GetClaimSize() string {
 
-func (g *gcePDCSIDriver) CreateDriver() {
+func (g *gcePDCSIDriver) CreateDriver(f *framework.Framework) (*testsuites.PerTestConfig, func()) {
+	framework.SkipIfMultizone(f.ClientSet)
```
what do you think about passing framework into SkipUnsupportedTest instead?

wongma7 replied:
nvm, the point of this is to avoid initializing framework

pohly replied:
nvm = never mind? If so, then I'll ignore the question ;-}

wongma7 reviewed (commented):
lgtm so far...IMO you can continue with other suites

davidz627 commented on test/e2e/storage/drivers/csi.go:
```diff
@@ -110,54 +107,56 @@ func (h *hostpathCSIDriver) GetClaimSize() string {
 	nodes := framework.GetReadySchedulableNodesOrDie(cs)
 	nodeName := nodes.Items[rand.Intn(len(nodes.Items))].Name
-	h.driverInfo.Config.ClientNodeName = nodeName
+	config := &testsuites.PerTestConfig{
```
its odd to me that `CreateDriver` would output a `PerTestConfig`. It may be cleaner to have whatever is calling `CreateDriver` to also call a `CreatePerTestConfig` as the next step and giving it the necessary inputs. 

`CreateDriver` can then just return whatever artifacts that only relate to the driver.

pohly replied:
`CreateDriver` always has been the "cluster is ready, set up test runtime resources" method. Going forward, drivers will be pre-installed during test cluster setup, but we might still need such a per-test setup function. Let's just rename `CreateDriver` to something that is a bit more generic. How about `PrepareTest`? What it does internally (install driver, deploy NFS server) is irrelevant at the API level, what matters is the result: cluster is ready, a per-test config describes it.

davidz627 replied:
got it. seems like the naming change will resolve this for me then

pohly replied:
Renamed.

davidz627 commented on test/e2e/storage/csi_volumes.go:
```diff
@@ -320,9 +227,18 @@ var _ = utils.SIGDescribe("CSI Volumes", func() {
+				cs := f.ClientSet
+				csics := f.CSIClientSet
+				ns := f.Namespace
+				driver := drivers.InitMockCSIDriver()
```
why do we Init a mock driver specifically here, then do a create CSI driver later?
Shouldn't the mock driver be one of many `csiDriver`'s permutations that we are programatically testing?

pohly replied:
`InitMockCSIDriver` doesn't initialize the driver. It just creates the TestDriver struct for it.

I don't know why the mock is treated special here.

davidz627 commented on test/e2e/storage/drivers/csi.go:
```diff
@@ -110,54 +107,56 @@ func (h *hostpathCSIDriver) GetClaimSize() string {
 		By(fmt.Sprintf("uninstalling %s driver", h.driverInfo.Name))
-		h.cleanup()
+		cleanup()
+		cancel()
```
should we be cancelling pod logging before or after cleanup? Also more descriptive names please. `cleanupDriver` and `cancelLogging`?

pohly replied:
okay

davidz627 commented on test/e2e/storage/drivers/csi.go:
```diff
@@ -310,8 +306,11 @@ func (g *gcePDCSIDriver) GetClaimSize() string {
 
-func (g *gcePDCSIDriver) CreateDriver() {
+func (g *gcePDCSIDriver) CreateDriver(f *framework.Framework) (*testsuites.PerTestConfig, func()) {
+	framework.SkipIfMultizone(f.ClientSet)
```
why did we move `SkipIfMultizone` from `SkipUnsupportedTest` to `CreateDriver`? It seems like it would belong in `SkipUnsupportedTest`

pohly replied:
Because `SkipUnsupportedTest` now runs before `framework.Framework` is initialized. Cheap tests that just check `e2e.test` parameters (like cloud provider) can and should be done in `SkipUnsupportedTest` to avoid setting up more expensive resources like `framework.Framework`. Tests that depend on a connection to the cluster can only be done once the framework is ready.

davidz627 replied:
lets document that entire statement in a comment so other dev's (probably a forgetful me) don't waste time trying to move that line back to `SkipUnsupportedTest` and having to figure out why it isn't working

pohly replied:
If you had attempted that, you would have found that you can't even get it compiled, because `SkipUnsupportedTest` doesn't have a `framework.Framework` pointer. But yes, let's document this better.

pohly replied:
Done.

davidz627 commented on test/e2e/storage/drivers/csi.go:
```diff
@@ -405,8 +403,12 @@ func (g *gcePDExternalCSIDriver) GetClaimSize() string {
-func (g *gcePDExternalCSIDriver) CreateDriver() {
-}
+func (g *gcePDExternalCSIDriver) CreateDriver(f *framework.Framework) (*testsuites.PerTestConfig, func()) {
+	framework.SkipIfMultizone(f.ClientSet)
```
same here, should be in `SkipUnsupportedTest`?

pohly replied:
See above.

davidz627 commented on test/e2e/storage/testsuites/driveroperations.go:
```diff
@@ -52,22 +52,6 @@ func CreateVolume(driver TestDriver, volType testpatterns.TestVolType) interface
 }
 
-// DeleteVolume deletes volume for test unless dynamicPV test
-func DeleteVolume(driver TestDriver, volType testpatterns.TestVolType, testResource interface{}) {
```
hard to figure out so i'm just going to ask. Why do we no longer need this `DeleteVolume` func?

pohly replied:
Because the new `TestVolume` interface has a `DeleteVolume` method that replaces the standalone `DeleteVolume` function. I had considered leaving an almost empty function like this in place for the sake of consistency:
```
func DeleteVolume(volume TestVolume) {
    if volume != nil {
        volume.DeleteVolume()
   }
}
```

But I decided against it because IMHO it just makes the overall code more complex. This is the same as opening a file: there is `f = open()` followed by ` f.Close` and not some `Close(f)`.

davidz627 commented on test/e2e/storage/testsuites/provisioning.go:
```diff
@@ -73,132 +76,88 @@ func (p *provisioningTestSuite) getTestSuiteInfo() TestSuiteInfo {
+		// Now do the more expensive test initialization.
+		config, driverCleanup = driver.CreateDriver(f)
+		claimSize := dDriver.GetClaimSize()
+		sc = dDriver.GetDynamicProvisionStorageClass(config, "")
```
at any point should we call the `driver.SkipUnsupportedTest` here?

pohly replied:
It was already called, see https://github.com/kubernetes/kubernetes/blob/4410815d24a4a25899bb40cb9744b0ba9a12bd5e/test/e2e/storage/testsuites/base.go#L96-L123

pohly replied:
That test runs before the testsuite specific code.

davidz627 commented on test/e2e/storage/testsuites/provisioning.go:
```diff
@@ -73,132 +76,88 @@ func (p *provisioningTestSuite) getTestSuiteInfo() TestSuiteInfo {
+		sc            *storage.StorageClass
+	)
+
+	init := func() {
```
nit: `init` could be more descriptive. `initTestCase`?

pohly replied:
okay

davidz627 commented on test/e2e/storage/testsuites/testdriver.go:
```diff
@@ -24,35 +24,44 @@ import (
+	// It sets up all necessary resources and returns the per-test configuration
+	// plus a cleanup function that frees all allocated resources.
+	CreateDriver(f *framework.Framework) (*PerTestConfig, func())
+	// SkipUnsupportedTest skips test if Testpattern is not suitable to test with the TestDriver
```
nit: add space

pohly replied:
okay

davidz627 reviewed (commented):
apologies for the late review. thanks for staying on top of this

pohly committed:
e2e: update bazel BUILD files

Generated via hack/update-bazel.sh.

pohly commented:
/hold

This is now a complete rewrite of all test suites. It compiles locally, but I'll check test results more carefully before asking for another round of reviews.

pohly commented:
/test pull-kubernetes-integration

pohly commented:
For pull-kubernetes-e2e-gce, the test numbers (755 passed, 2255 skipped) in https://gubernator.k8s.io/build/kubernetes-jenkins/pr-logs/pull/72434/pull-kubernetes-e2e-gce/68530/ are the same as in another recent build against some other PR (https://gubernator.k8s.io/build/kubernetes-jenkins/pr-logs/pull/74028/pull-kubernetes-e2e-gce/68534/). A diff confirms that also all test names are the same.

pohly committed:
e2e/storage: remove test pattern filtering

The recommended approach for not running unsuitable tests is to skip
them at runtime with an explanation. Filtering out unsuitable test
patters and thus not even defining unsuitable tests was done earlier
because it was faster than skipping tests at runtime.

But now these tests can be skipped efficiently, so this special case
can be removed.

pohly commented:
/hold cancel

@wongma7 if test results become green, can you have another look at this PR? They were okay before I added 8a4dd7f91b83034fe1b7d59c68388323b9e2e8ec (, so everything before that should be done. I don't know whether that last commit broke something (it might, if some tests aren't get skipped properly) and I can't wait longer today for the testing to complete.

I'd like to get this merged relatively fast, because of the inevitable conflicts when others work on these tests in parallel.

msau42 commented:
/test pull-kubernetes-e2e-gce-alpha-features
/test pull-kubernetes-e2e-gce-csi-serial

pohly commented:
/test pull-kubernetes-e2e-gce-alpha-features
/test pull-kubernetes-e2e-gce-csi-serial

pohly commented:
`[sig-storage] CSI Volumes CSI attach test using mock driver [Feature:CSIDriverRegistry] should not require VolumeAttach for drivers without attachment` failed in https://gubernator.k8s.io/build/kubernetes-jenkins/pr-logs/pull/72434/pull-kubernetes-e2e-gce-alpha-features/108/, but that same error also occurred while working on https://github.com/kubernetes/kubernetes/pull/73677 (see https://gubernator.k8s.io/build/kubernetes-jenkins/pr-logs/pull/73677/pull-kubernetes-e2e-gce-alpha-features/104/), so I'll assume that this isn't caused by my changes.

pohly commented:
/test pull-kubernetes-e2e-gce-alpha-features
/test pull-kubernetes-e2e-gce-csi-serial

pohly commented:
@msau42 @wongma7 @davidz627: all tests green (including gce alpha and serial), no further changes planned as far as I am concerned -> please have a look.

wongma7 commented on test/e2e/storage/testsuites/volume_io.go:
```diff
@@ -74,87 +74,59 @@ func (t *volumeIOTestSuite) getTestSuiteInfo() TestSuiteInfo {
+		// Now do the more expensive test initialization.
+		config, testCleanup = driver.PrepareTest(f)
+		resource = createGenericVolumeTestResource(driver, config, pattern)
+		if resource.volSource == nil {
```
nit, check not needed

pohly replied:
Probably not, but it was there in the original code and I wanted to avoid changing test logic.

wongma7 commented on test/e2e/storage/testsuites/volumes.go:
```diff
@@ -89,101 +89,76 @@ func skipExecTest(driver TestDriver) {
+		// Now do the more expensive test initialization.
+		config, testCleanup = driver.PrepareTest(f)
+		resource = createGenericVolumeTestResource(driver, config, pattern)
+		if resource.volSource == nil {
```
same

pohly replied:
Same :man_shrugging:. I wanted to avoid changing test logic.

wongma7 commented on test/e2e/storage/testsuites/provisioning.go:
```diff
@@ -76,183 +79,156 @@ func (p *provisioningTestSuite) getTestSuiteInfo() TestSuiteInfo {
+		claimSize := dDriver.GetClaimSize()
+		sc = dDriver.GetDynamicProvisionStorageClass(config, "")
+		if sc == nil {
+			framework.Skipf("Driver %q does not define Dynamic Provision StorageClass - skipping", dInfo.Name)
```
these Skip messages should specify that the sc is not defined because the fstype %s is not supported. there is no other reason (currently?) for a driver to implement DynamicPVTestDriver but return nil

pohly replied:
Can this code here know why the driver didn't return a storage class? I don't think it should make assumptions based on the current implementation.

wongma7 replied:
yeah I guess it's fine as-is. I guess it is the driver author's responsibility to make sense of why their GetDynamicProvisionStorageClass would return nil. Maybe it could return an error in the future, seems pointless now though

wongma7 commented on test/e2e/storage/testsuites/provisioning.go:
```diff
@@ -283,18 +259,22 @@ func testProvisioning(input *provisioningTestInput) {
+// TestDynamicProvisioning tests dynamic provisioning with specified StorageClassTest
+func (t StorageClassTest) TestDynamicProvisioning() *v1.PersistentVolume {
+	client := t.Client
+	claim := t.Claim
```
could fail w/ nice error if this is nil

pohly replied:
Not sure if Github is showing me the right context. What can be nil here? `t.Client` and/or `t.Claim`?

Are you worried that they might be accidentally passed as nil because they are now struct members instead of separate function parameters? I agree, this isn't caught explicitly. If this is the only requested change, then I'd prefer to add assertions in a separate PR.

wongma7 replied:
I meant t.Claim, though client could be nil as well. either way, the nil error will be caught a few lines below, so yeah these assertions can be added in a separate PR as "nice-to-haves" :+1:

pohly replied:
As I was rebasing anyway, I added the asserts.

wongma7 reviewed (approved):
I just had some inconsequential suggestions.
driver & driver interface changes lgtm.
testsuite changes largely lgtm, may need to do another pass. Agree with removing skipUnsupportedTest from testsuites, only snapshots implemented it

jarrpa reviewed (approved):
I can't speak to the full PR, but the parts I'm familiar with LGTM.

davidz627 commented on test/e2e/storage/csi_volumes.go:
```diff
@@ -343,27 +265,28 @@ var _ = utils.SIGDescribe("CSI Volumes", func() {
-					defer destroyCSIDriver(csics, driver)
-					defer driver.CleanupDriver()
+					defer destroyCSIDriver(csics, config.GetUniqueDriverName())
+					defer testCleanup()
```
shouldn't we want to cleanup the test whether the driver exists or not

pohly replied:
Yes. This was introduced in https://github.com/kubernetes/kubernetes/pull/74079 and I noticed it too when rebasing on top of that, then notified @jarrpa. There's now a fix pending in https://github.com/kubernetes/kubernetes/pull/74079... which has been merged. I'll need to rebase once more.

davidz627 commented on test/e2e/storage/testsuites/subpath.go:
```diff
@@ -71,345 +72,351 @@ func (s *subPathTestSuite) getTestSuiteInfo() TestSuiteInfo {
+			By("Deleting pod")
+			err := framework.DeletePodWithWait(f, cs, pod)
+			Expect(err).ToNot(HaveOccurred(), "while deleting pod")
+			pod = nil
```
this is suspicious to me. I want to avoid a pattern where we are reusing structs and just clearing field. Is there a way we can start from scratch instead?

pohly replied:
The test does start from scratch, with a fresh copy of all local variables. Setting the pointers to nil is meant to ensure that `cleanup` can be called more than once, should the need ever arise. Currently this isn't done, so I could remove the `pod = nil` and leave the pointers to already deleted items, but personally would prefer to keep this extra precaution.

pohly replied:
Hmm, I need to be more precise: each test starts with a fresh copy of the the local variables *if* `init` initializes all of them. I see how that might be something that can be missed for a complex test suite where some variables are only set under certain conditions.

This can be avoided by storing all these local variables in a struct and re-initializing the struct before setting individual members. I've done that in commit ec3655a1d40ced6b1873e627b736aae1cf242477.

I'm undecided whether it makes the code easier to read ("shared" variables clearly marked with `l.` prefix) or harder to read (extra prefix). @davidz627 Please let me know whether we should do it this way or by carefully setting all variables in init.

davidz627 reviewed (commented):
mostly LGTM. Just a couple small comments. Thanks for doing this!

pohly committed:
e2e/storage: re-initialize all local variables

There is a risk that the init function does not reset one of the local
variables that was set by a previous test. To avoid this, all
variables set by init are now in a struct which gets cleaned
completely first.

pohly commented:
/test pull-kubernetes-e2e-gce-alpha-features
/test pull-kubernetes-e2e-gce-csi-serial

pohly commented:
/priority important-soon

msau42 commented:
@childsb this should be 1.14 milestone right?

/approve

will let @davidz627 and/or @wongma7 lgtm

davidz627 commented:
/lgtm
